C7Engine/Components/ComponentManager.cs

```diff
@@ -0,0 +1,57 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+// modeled after the Unity Toolbox pattern from https://wiki.unity3d.com/index.php/Toolbox
+namespace C7Engine.Components
+{
+    public sealed class ComponentManager
+    {
+        // singleton implemenation taken from example at https://csharpindepth.com/Articles/Singleton
+        private static readonly ComponentManager _instance = new ComponentManager();
+
+        static ComponentManager()
+        {
+
+        }
+
+        private ComponentManager()
+        {
+
+        }
+
+        public static ComponentManager Instance
+        {
+            get { return _instance; }
+        }
+
+        // type dictionary taken from Jon Skeet's implementation at https://codeblog.jonskeet.uk/2008/10/08/mapping-from-a-type-to-an-instance-of-that-type/
+        private Dictionary<Type, GameComponent> _components = new Dictionary<Type, GameComponent>();
+
+        public ComponentManager AddComponent<T>(T component) where T : GameComponent
```

Comment:
if you try to add a type that's already in the dictionary, what's the behavior? Does it get overwritten, is it a no-op, does it throw an exception, etc.?